clash-prelude/tests/Clash/Tests/AsyncFIFOSynchronizer.hs

```diff
@@ -963,6 +970,149 @@ test6W7 = testW sync6W7
   , (Nothing,False)
   ]
 
+createDomain vSystem{vName="A", vPeriod=hzToPeriod 20e6} -- fast
+createDomain vSystem{vName="B", vPeriod=hzToPeriod 10e6} -- slow
+createDomain vSystem{vName="C", vPeriod=hzToPeriod 7e6} -- slower
+
+data DomProxy (dom :: Domain) where
+  DomProxy
+    :: KnownDomain dom
+    => DomProxy dom
+
+-- A more useful Show instance than the one for 'Proxy'
+instance Show (DomProxy dom) where
+  showsPrec d dom@DomProxy =
+    showParen (d > app_prec) $ ("DomProxy @" <>) . (symbolVal dom <>)
+   where app_prec = 10
+
+data NamedTest a = NamedTest
+  { namedTest :: a
+  , testName :: String
+  }
+
+instance Show (NamedTest a) where
+  show nt = testName nt
+
+testNameShowsPrec
+  :: Int
+  -> ShowS
+  -> DomProxy dom1
+  -> DomProxy dom2
+  -> ShowS
+testNameShowsPrec d baseName dom1 dom2 =
+  showParen (d > app_prec) $ baseName . (' ':) . showsPrec 11 dom1 .
+    (' ':) . showsPrec 11 dom2
+ where app_prec = 10
+
+forAllNamedTestProperties
+  :: [NamedTest (PropertyT IO ())]
+  -> Property
+
+forAllNamedTestProperties nts = property $ do
+  t <- fmap namedTest . forAllWith show . Gen.resize 99 $ Gen.element nts
+  t
+
+data WriteAction =
+    Write
+  | WNoOp
+  | WGated
+  | GatedWrite
+  deriving (Eq, Enum, Bounded, Show, Generic, NFDataX)
+
+data ReadAction =
+    Read
+  | RNoOp
+  | RGated
+  | GatedRead
+  deriving (Eq, Enum, Bounded, Show, Generic, NFDataX)
+
+fifoOperations
+  :: forall wdom rdom
+   . (KnownDomain wdom, KnownDomain rdom)
+  => [ReadAction]
+  -> [WriteAction]
+  -> ( Signal rdom ( Bool
+                     -- Test fully done
+                   , Maybe Int
+                     -- Element read from FIFO
+                   )
+     , Signal wdom ( Bool
+                     -- Test fully done
+                   , Maybe Int
+                    -- Element written to FIFO
+                   )
+     )
+fifoOperations racts wacts = (bundle (rAllDone, rel), bundle (wAllDone, wel))
+ where
+  (rdata, rempty, wfull) =
+    asyncFIFOSynchronizer d3 wclk rclk noWRst noRRst (toEnable wen)
+      (toEnable ren) rinc wDataM
+  rclk = clockGen @rdom
+  wclk = clockGen @wdom
+  noRRst = unsafeFromHighPolarity @rdom (pure False)
+  noWRst = unsafeFromHighPolarity @wdom (pure False)
```

Comment:
Does it make sense to not reset at all? If so, could you add a comment why?

Reply:
Good question, I don't really know. I always assumed we do it to get rid of undefined startup values - even though we don't actually test with domains that don't have defined startup values :|. My question was triggered by it being different from all the other testbenches. 

> Also, Xilinx even recommends avoiding resets in FPGA's if possible.

That's true, but they do recommend designing it in such a way that you can still reset your system. I.e., a pipeline might use resetless registers, but should nonetheless still converge to all known values while asserting the reset by virtue of registers in front of the pipeline.

I agree with your general thesis, we don't typically test "interesting" properties of the reset. I believe this to be an artifact of the painfulness that is writing Clash HDL testbenches..

Reply:
My `asyncRam` test benches also don't reset, and I changed the TDP BRAM test benches to no longer reset as the logic that was there to handle the first two samples was odd. So it's not the first instance, although those tests are recent!